### PR TITLE
fix #1895

### LINKED ISSUE #1895: Respect `Prefer` header with validation proxy when server returns 501
## Summary

I was able to use `Prefer` header to control returned mock response when using `mock-server` (`Prefer: example=example-key`). Then I switched to `validation-proxy` , hard-coded my upstream sever to return 501 for all requests and send a request to the `validation-proxy` with `Prefer` header as I did with `mock-server`. Surprisingly, it didn't work. I always returned the first example.

Is this how the `validation-proxy` is supposed to work?

## Additional context

Prism: 4.4.0 (get via docker)

### PR DESCRIPTION
**Pull request by nursanamar**
Addresses #1895

**Summary**

**Checklist**

- The basics
  - [X] I tested these changes manually in my local or dev environment
- Tests
  - [X] N/A

### CODE REVIEW

File: packages/core/src/factory.ts [hunk 1]
```diff
@@ -62,11 +63,23 @@ export function factory<Resource, Input, Output, Config extends IPrismConfig>(
     config: Config,
     validations: IPrismDiagnostic[]
   ): TE.TaskEither<Error, ResourceAndValidation & { output: Output }> => {
+    
+    const input = {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
Comment (daniel-white) [inline packages/core/src/factory.ts:68]: what typescript errors are these hiding? is there another way to express what you need this to?

File: packages/http-server/src/getHttpConfigFromRequest.ts [hunk 2]
```diff
@@ -21,13 +21,21 @@ const PreferencesDecoder = D.partial({
 
 type RequestPreferences = Partial<Omit<IHttpOperationConfig, 'mediaType'>>;
 
-export const getHttpConfigFromRequest = (
+export const getRequestPrefrences = (
```
[the remark this answers is not in the document]
Comment (daniel-white) [reply packages/http-server/src/getHttpConfigFromRequest.ts:24]: also what is this doing other than adding an extra function?

**Comment by nursanamar** [comment]
@daniel-white thank you for the reviews, about the test I don't really sure how to test this fix.

**Comment by nursanamar** [comment]
Add some tests

File: packages/core/src/factory.ts [hunk 3]
```diff
@@ -6,8 +6,10 @@ import { pipe } from 'fp-ts/function';
 import { defaults } from 'lodash';
 import { IPrism, IPrismComponents, IPrismConfig, IPrismDiagnostic, IPrismProxyConfig, IPrismOutput } from './types';
 import { getSemigroup, NonEmptyArray } from 'fp-ts/NonEmptyArray';
-import { DiagnosticSeverity } from '@stoplight/types';
+import { DiagnosticSeverity, IHttpRequest } from '@stoplight/types';
+import { IHttpNameValue } from "@stoplight/prism-http/src/types";
 import { identity } from 'fp-ts/function';
+import { getHttpConfigFromRequest } from '@stoplight/prism-http-server/src/getHttpConfigFromRequest';
```
Comment (daniel-white) [inline packages/core/src/factory.ts:12]: we do not want a circular dependency between core and http. could we change the call sites to `mockOrForward` to include calculation of the mocking config?

File: packages/core/src/factory.ts [hunk 4]
```diff
@@ -62,12 +64,30 @@ export function factory<Resource, Input, Output, Config extends IPrismConfig>(
     config: Config,
     validations: IPrismDiagnostic[]
   ): TE.TaskEither<Error, ResourceAndValidation & { output: Output }> => {
-    const mockCall = () =>
-      components.mock({
+    const mockCall = () => {
```
Comment (daniel-white) [inline packages/core/src/factory.ts:67]: in the event `mockOrForward`'s call sites include the mock config, this should be all reverted

File: packages/core/src/factory.ts [hunk 5]
```diff
@@ -62,12 +64,30 @@ export function factory<Resource, Input, Output, Config extends IPrismConfig>(
...
+
+      // Construct input
+      const input = {
+        url: {
```
Comment (daniel-white) [inline packages/core/src/factory.ts:71]: this also should support the `__prefer` query string options as well.

**Review by daniel-white — CHANGES_REQUESTED** [review]

**Pull request closed without merging**